src/main/java/com/conveyal/analysis/results/CsvResultWriter.java

```diff
@@ -2,76 +2,102 @@
+    /** Override to check the size of incoming results and ensure they match expected dimensions. */
+    protected abstract void checkDimension (RegionalWorkResult workResult);
 
     /**
      * Construct a writer to record incoming results in a CSV file, with header row consisting of
      * "origin", "destination", and the supplied indicator.
      * FIXME it's strange we're manually passing injectable components into objects not wired up at application construction.
      */
-    CsvResultWriter (RegionalTask task, String outputBucket, FileStorage fileStorage, Result resultType) throws IOException {
+    CsvResultWriter (RegionalTask task, String outputBucket, FileStorage fileStorage) throws IOException {
         super(fileStorage);
         super.prepare(task.jobId, outputBucket);
-        this.resultType = resultType;
-        this.fileName = task.jobId + "_" + resultType +".csv";
+        this.fileName = task.jobId + "_" + resultType() +".csv";
         BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(bufferFile));
         csvWriter = new CsvWriter(bufferedWriter, ',');
+        setDataColumns(columnHeaders());
+        this.task = task;
+        LOG.info("Created CSV file to hold {} results for regional job {}", resultType(), task.jobId);
     }
 
     /**
-     * Writes a header row, including "origin," "destination," and the supplied data columns
+     * Writes a header row containing the supplied data columns.
      */
-    public void setDataColumns(String... columns) throws IOException {
+    protected void setDataColumns(String... columns) throws IOException {
         this.nDataColumns = columns.length;
-        csvWriter.writeRecord(ArrayUtils.addAll(new String[]{"origin", "destination"}, columns));
-        LOG.info("Created csv file to store {} results from workers.", resultType);
+        csvWriter.writeRecord(columns);
     }
 
     /**
-     * Gzip the csv file and upload it to S3.
+     * Gzip the csv file and move it into permanent file storage such as AWS S3.
+     * Note: stored file will undergo gzip compression in super.finish(), but be stored with a .csv extension.
+     * When this file is downloaded from the UI, the browser will decompress, yielding a logically named .csv file.
+     * Downloads through another channel (e.g. aws s3 cp), will need to be decompressed manually.
      */
     protected synchronized void finish () throws IOException {
         csvWriter.close();
-        // Note that stored file undergo gzip compression in super.finish(), but be stored with a .csv extension.
-        // When this file is downloaded from the UI, the browser will decompress, yielding a logically named .csv file.
-        // Downloads through another channel (e.g. aws s3 cp), will need to be decompressed manually.
         super.finish(this.fileName);
     }
 
     /**
-     * Write a single row into the CSV file.
+     * Write all rows for a single regional work result (single origin) into the CSV file.
```

Comment:
Good correction - the CSV is sometimes expected to have multiple rows per origin (e.g. paths)